src/headers.rs

```diff
@@ -0,0 +1,39 @@
+//! Support library for handling headers in a safe manner
+
+use hyper::header::{Header, Headers};
+
+/// Trait to add a mechanism to safely retrieve a header.
+///
+/// In Hyper 0.11, if you add an Authorization<Basic> header,
+/// and then attempt get an Authorization<Bearer> header, the code
+/// will panic, as the type ID doesn't match.
+pub trait SafeHeaders {
+    /// Safely get a header from a hyper::header::Headers
+    fn safe_get<H: Header>(&self) -> Option<H>;
+}
+
+impl SafeHeaders for Headers {
+    fn safe_get<H: Header>(&self) -> Option<H> {
+        self.get_raw(H::header_name())
+            .map(H::parse_header)
+            .map(Result::ok)
+            .unwrap_or(None)
```

Comment:
Isn't this `unwrap_or` pointless?  `Result::ok` turns error types into `None`?

Reply:
`get_raw()` returns an Option<&Raw>.

After we've mapped this through `H::parse_header`, we get `Option<Result<H>>`

We don't care about the error here, so we discard this with `Result::ok` giving us `Option<Option<H>>`, but we want to return `Option<H>`.

There's `.flatten()` in the stdlib for this, but it's not stable, so instead we do `.unwap_or(None)` to flatten the Option.